compiler/src/dotty/tools/dotc/transform/init/Potentials.scala

```diff
@@ -0,0 +1,218 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import scala.collection.mutable
+
+import ast.tpd._
+import reporting.trace
+import config.Printers.init
+
+import core._
+import Types._, Symbols._, Contexts._
+
+import Effects._, Summary._
+
+object Potentials {
+  type Potentials = Set[Potential]
+  val empty: Potentials = Set.empty
+
+  def show(pots: Potentials)(implicit ctx: Context): String =
+    pots.map(_.show).mkString(", ")
+
+  /** A potential represents an aliasing of a value that is possibly under initialization */
+  sealed trait Potential {
+    def size: Int
+    def show(implicit ctx: Context): String
+    def source: Tree
+  }
+
+  /** The object pointed by `C.this` */
+  case class ThisRef(classSymbol: ClassSymbol)(val source: Tree) extends Potential {
+    val size: Int = 1
+    def show(implicit ctx: Context): String = classSymbol.name.show + ".this"
+
+    /** Effects of a method call or a lazy val access
+     *
+     *  It assumes all the outer `this` are fully initialized.
+     */
+    def effectsOf(sym: Symbol)(implicit env: Env): Effects = trace("effects of " + sym.show, init, r => Effects.show(r.asInstanceOf)) {
+      val cls = sym.owner.asClass
+      val effs = env.summaryOf(cls).effectsOf(sym)
+      Effects.asSeenFrom(effs, this, cls, Potentials.empty)
+    }
+
+    /** Potentials of a field, a method call or a lazy val access
+     *
+     */
+    def potentialsOf(sym: Symbol)(implicit env: Env): Potentials = trace("potentials of " + sym.show, init, r => Potentials.show(r.asInstanceOf)) {
+      val cls = sym.owner.asClass
+      val pots = env.summaryOf(cls).potentialsOf(sym)
+      Potentials.asSeenFrom(pots, this, cls, Potentials.empty)
+    }
+  }
+
+  /** The object pointed by `C.super.this`, mainly used for override resolution */
+  case class SuperRef(pot: Potential, supercls: ClassSymbol)(val source: Tree) extends Potential {
+    val size: Int = 1
+    def show(implicit ctx: Context): String = pot.show + ".super[" + supercls.name.show + "]"
+  }
+
+  /** A warm potential represents an object of which all fields are initialized, but it may contain
+   *  reference to objects under initialization.
+   *
+   *  @param classSymbol  The concrete class of the object
+   *  @param outer        The potential for `this` of the enclosing class
+   */
+  case class Warm(classSymbol: ClassSymbol, outer: Potential)(val source: Tree) extends Potential {
+    def size: Int = 1
+    def show(implicit ctx: Context): String = "Warm[" + classSymbol.show + ", outer = " + outer.show + "]"
+
+    /** Effects of a method call or a lazy val access
+     *
+     *  The method performs prefix and outer substitution
+     */
+    def effectsOf(sym: Symbol)(implicit env: Env): Effects = trace("effects of " + sym.show, init, r => Effects.show(r.asInstanceOf)) {
+      val cls = sym.owner.asClass
```

Comment:
You could define `cls` and `effs` as `protected def`s in `trait Potential`. This would DRY the code here and in `potentialsOf`. Also it would DRY the same two methods in `ThisRef`.

Reply:
`cls` and `effs` do not make much sense in `trait Potential `, I feel a little duplication here is better as it avoids unnecessary contracts & coupling.